File: code/27.cpp
class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int n = nums.size();
        if (n == 0) {
            return 0;
        }
        int i = 0;
        int j = n - 1;
        int k = 0;
        while (nums[j] == val) {
            j--;
            if (j < 0) {
                break;
            }
        }
        while (i <= j) {
            if (nums[i] != val) {
                k++;
                i++;
            }
            else {
                // Replace with one of the non-vals at the end of the array
                nums[i] = nums[j];
                j--;
                while (nums[j] == val) {
                    j--;
                    if (j < i) {
                        break;
                    }
                }
            }
        }
        return k;
    }
};

File: code/27.cs
public class Solution {
    public int RemoveElement(int[] nums, int val) {
        int changed = 0;
        int i = 0;
        int n = nums.Length;
        while (i < n) {
            while (n > 0 && nums[n-1] == val) {
                n--;
            }
            if (i >= n) {
                break;
            }
            if (nums[i] == val) {
                changed++;
                int a = nums[i];
                int b = nums[n-1];
                nums[i] = b;
                nums[n-1] = a;
            }
            i++;
        }
        return n;

    }
}
